File: components/espectre/p95_calibrator.cpp
/*
 * ESPectre - P95 Calibrator Implementation
 * 
 * P95-based band selection algorithm for optimal subcarrier selection.
 * Inherits common lifecycle from BaseCalibrator.
 * 
 * Author: Francesco Pace <francesco.pace@gmail.com>
 * License: GPLv3
 */

#include "p95_calibrator.h"
#include "threshold.h"
#include "csi_manager.h"
#include "utils.h"
#include "esphome/core/log.h"
#include <algorithm>
#include <cmath>
#include <cstring>
#include <limits>
#include "freertos/FreeRTOS.h"
#include "freertos/task.h"

namespace esphome {
namespace espectre {

static const char *TAG = "P95Calibrator";

// ============================================================================
// PUBLIC API
// ============================================================================

void P95Calibrator::init_subcarrier_config() {
  ESP_LOGI(TAG, "P95: HT20 mode, %d subcarriers, valid range [%d-%d], DC=%d",
           HT20_NUM_SUBCARRIERS, HT20_GUARD_BAND_LOW, HT20_GUARD_BAND_HIGH, HT20_DC_SUBCARRIER);
}

// ============================================================================
// CALIBRATION ALGORITHM
// ============================================================================

esp_err_t P95Calibrator::run_calibration_() {
  uint16_t buffer_count = file_buffer_.get_count();
  
  if (buffer_count < MVS_WINDOW_SIZE + 10) {
    ESP_LOGE(TAG, "Not enough packets for calibration (%d < %d)", 
             buffer_count, MVS_WINDOW_SIZE + 10);
    return ESP_FAIL;
  }
  
  ESP_LOGD(TAG, "Starting P95-based band calibration...");
  ESP_LOGD(TAG, "  Buffer: %d packets, MVS window: %d", buffer_count, MVS_WINDOW_SIZE);
  
  // Read all packets from file
  std::vector<uint8_t> all_data = file_buffer_.read_window(0, buffer_count);
  if (all_data.size() != buffer_count * HT20_NUM_SUBCARRIERS) {
    ESP_LOGE(TAG, "Failed to read calibration data");
    return ESP_FAIL;
  }
  
  // If skipping subcarrier selection, just calculate baseline
  if (skip_subcarrier_selection_) {
    selected_band_size_ = current_band_.size();
    std::memcpy(selected_band_, current_band_.data(), selected_band_size_);
    
    BandResult result = evaluate_band_(all_data, selected_band_, selected_band_size_);
    mv_values_ = std::move(result.mv_values);
    
    ESP_LOGI(TAG, "Baseline calibration complete (fixed subcarriers)");
    ESP_LOGD(TAG, "  P95: %.4f", result.p95);
    
    return ESP_OK;
  }
  
  // Step 1: Generate all candidate bands (12 consecutive subcarriers)
  std::vector<std::vector<uint8_t>> candidates;
  get_candidate_bands_(candidates);
  
  if (candidates.empty()) {
    ESP_LOGE(TAG, "No valid candidate bands found");
    return ESP_FAIL;
  }
  
  ESP_LOGI(TAG, "Evaluating %zu candidate bands...", candidates.size());
  
  // Step 2: Evaluate each candidate band
  std::vector<BandResult> results;
  results.reserve(candidates.size());
  
  for (size_t i = 0; i < candidates.size(); i++) {
    BandResult result = evaluate_band_(all_data, candidates[i].data(), 
                                       static_cast<uint8_t>(candidates[i].size()));
    result.start_sc = candidates[i][0];
    results.push_back(std::move(result));
    
    ESP_LOGV(TAG, "  Band [%d-%d]: P95=%.4f, FP=%.1f%%",
             candidates[i][0], candidates[i].back(), results.back().p95, results.back().fp_estimate * 100.0f);
    
    if (i % 10 == 0) {
      vTaskDelay(1);
    }
  }
  
  // Step 3: Select optimal band
  float p95_limit = MVS_THRESHOLD - SAFE_MARGIN;
  
  size_t best_idx = 0;
  float best_p95 = -1.0f;
  bool found_safe = false;
  
  for (size_t i = 0; i < results.size(); i++) {
    if (results[i].p95 < p95_limit) {
      if (results[i].p95 > best_p95) {
        best_p95 = results[i].p95;
        best_idx = i;
        found_safe = true;
      }
    }
  }
  
  if (!found_safe) {
    ESP_LOGW(TAG, "No bands with P95 < %.2f, using lowest", p95_limit);
    best_p95 = std::numeric_limits<float>::infinity();
    for (size_t i = 0; i < results.size(); i++) {
      if (results[i].p95 < best_p95) {
        best_p95 = results[i].p95;
        best_idx = i;
      }
    }
  } else {
    ESP_LOGD(TAG, "Found safe bands (P95 < %.2f)", p95_limit);
  }
  
  // Copy selected band and mv_values
  std::memcpy(selected_band_, candidates[best_idx].data(), HT20_SELECTED_BAND_SIZE);
  selected_band_size_ = HT20_SELECTED_BAND_SIZE;
  mv_values_ = std::move(results[best_idx].mv_values);
  
  ESP_LOGI(TAG, "P95: Band selection successful");
  ESP_LOGD(TAG, "  Selected: [%d-%d]",
           selected_band_[0], selected_band_[HT20_SELECTED_BAND_SIZE - 1]);
  ESP_LOGD(TAG, "  P95 MV: %.4f", best_p95);
  ESP_LOGD(TAG, "  Est. FP rate: %.1f%%", results[best_idx].fp_estimate * 100.0f);
  
  return ESP_OK;
}

void P95Calibrator::get_candidate_bands_(std::vector<std::vector<uint8_t>>& candidates) {
  candidates.clear();
  
  // HT20: Generate valid bands of 12 consecutive subcarriers
  // Zone before DC (subcarriers 11-31)
  for (uint16_t start = HT20_GUARD_BAND_LOW; start + HT20_SELECTED_BAND_SIZE <= HT20_DC_SUBCARRIER; start++) {
    std::vector<uint8_t> band;
    bool valid = true;
    
    for (uint8_t i = 0; i < HT20_SELECTED_BAND_SIZE; i++) {
      uint8_t sc = start + i;
      if (sc < HT20_GUARD_BAND_LOW || sc > HT20_GUARD_BAND_HIGH || sc == HT20_DC_SUBCARRIER) {
        valid = false;
        break;
      }
      band.push_back(sc);
    }
    
    if (valid && band.size() == HT20_SELECTED_BAND_SIZE) {
      candidates.push_back(band);
    }
  }
  
  // Zone after DC (subcarriers 33-52)
  for (uint16_t start = HT20_DC_SUBCARRIER + 1; start + HT20_SELECTED_BAND_SIZE <= HT20_GUARD_BAND_HIGH + 1; start++) {
    std::vector<uint8_t> band;
    bool valid = true;
    
    for (uint8_t i = 0; i < HT20_SELECTED_BAND_SIZE; i++) {
      uint8_t sc = start + i;
      if (sc < HT20_GUARD_BAND_LOW || sc > HT20_GUARD_BAND_HIGH || sc == HT20_DC_SUBCARRIER) {
        valid = false;
        break;
      }
      band.push_back(sc);
    }
    
    if (valid && band.size() == HT20_SELECTED_BAND_SIZE) {
      candidates.push_back(band);
    }
  }
  
  ESP_LOGD(TAG, "Generated %zu candidate bands (HT20)", candidates.size());
}

P95Calibrator::BandResult P95Calibrator::evaluate_band_(
    const std::vector<uint8_t>& all_data, 
    const uint8_t* band, 
    uint8_t band_size) {
  
  BandResult result = {0, std::numeric_limits<float>::infinity(), 1.0f, {}};
  
  if (all_data.size() < MVS_WINDOW_SIZE * HT20_NUM_SUBCARRIERS) {
    return result;
  }
  
  uint16_t num_packets = all_data.size() / HT20_NUM_SUBCARRIERS;
  
  // Calculate turbulence for each packet
  std::vector<float> turbulences(num_packets);
  float valid_mags[HT20_SELECTED_BAND_SIZE];
  
  for (uint16_t pkt = 0; pkt < num_packets; pkt++) {
    const uint8_t* packet_data = &all_data[pkt * HT20_NUM_SUBCARRIERS];
    
    for (uint8_t i = 0; i < band_size && i < HT20_SELECTED_BAND_SIZE; i++) {
      valid_mags[i] = static_cast<float>(packet_data[band[i]]);
    }
    
    turbulences[pkt] = std::sqrt(calculate_variance_two_pass(valid_mags, band_size));
  }
  
  // Calculate moving variance series
  result.mv_values.reserve(num_packets - MVS_WINDOW_SIZE);
  
  for (uint16_t i = MVS_WINDOW_SIZE; i < num_packets; i++) {
    float variance = calculate_variance_two_pass(&turbulences[i - MVS_WINDOW_SIZE], MVS_WINDOW_SIZE);
    result.mv_values.push_back(variance);
  }
  
  if (result.mv_values.empty()) {
    return result;
  }
  
  // Calculate P95 for band selection
  result.p95 = calculate_percentile(result.mv_values, BAND_SELECTION_PERCENTILE);
  
  // Estimate FP rate
  uint32_t fp_count = 0;
  for (float v : result.mv_values) {
    if (v > MVS_THRESHOLD) {
      fp_count++;
    }
  }
  result.fp_estimate = static_cast<float>(fp_count) / result.mv_values.size();
  
  return result;
}


}  // namespace espectre
}  // namespace esphome

File: micro-espectre/src/p95_calibrator.py
"""
P95 Calibrator - Automatic Subcarrier Band Selection

Selects optimal 12-subcarrier band for motion detection by minimizing
the P95 of moving variance during baseline. This directly optimizes
for low false positive rate.

Algorithm (P95 Band Selection):
1. Collect baseline CSI packets (quiet room)
2. For each candidate band of 12 consecutive subcarriers:
   - Calculate moving variance series
   - Compute P95 of moving variance for band ranking
3. Select the band with LOWEST P95 (furthest from threshold)

Returns (band, mv_values). Adaptive threshold is calculated externally
using threshold.py after band selection.

Author: Francesco Pace <francesco.pace@gmail.com>
License: GPLv3
"""

import math
import gc
import os

# Import HT20 constants from config
from src.config import (
    NUM_SUBCARRIERS, EXPECTED_CSI_LEN,
    GUARD_BAND_LOW, GUARD_BAND_HIGH, DC_SUBCARRIER, BAND_SIZE
)

# Import utility functions
from src.utils import (
    calculate_percentile,
    calculate_spatial_turbulence,
    calculate_moving_variance
)

# Import interface
from src.calibrator_interface import ICalibrator

# Constants
BUFFER_FILE = '/p95_buffer.bin'

# Threshold for null subcarrier detection
NULL_SUBCARRIER_THRESHOLD = 1.0

# MVS parameters
MVS_WINDOW_SIZE = 50
MVS_THRESHOLD = 1.0

# Band selection uses fixed P95 (algorithm constant)
BAND_SELECTION_PERCENTILE = 95


def cleanup_buffer_file():
    """Remove any leftover buffer file from previous interrupted runs."""
    try:
        os.remove(BUFFER_FILE)
        print("P95: Cleaned up leftover buffer file")
    except OSError:
        pass


class P95Calibrator(ICalibrator):
    """
    Automatic band calibrator using P95 moving variance optimization.
    
    Selects the optimal 12-subcarrier band by finding the one with
    lowest P95 of moving variance during baseline (minimizes FP rate).
    
    Uses file-based storage to avoid RAM limitations on ESP32.
    HT20 only: 64 subcarriers.
    """
    
    def __init__(self, buffer_size=700):
        """
        Initialize P95 calibrator.
        
        Args:
            buffer_size: Number of packets to collect (default: 700)
        """
        self.buffer_size = buffer_size
        self._packet_count = 0
        self._filtered_count = 0
        self._file = None
        self._initialized = False
        
        # Remove old buffer file
        try:
            os.remove(BUFFER_FILE)
        except OSError:
            pass
        
        self._file = open(BUFFER_FILE, 'wb')
    
    def add_packet(self, csi_data):
        """
        Add CSI packet to calibration buffer.
        
        HT20 only: expects 128 bytes (64 subcarriers x 2 I/Q).
        
        Args:
            csi_data: CSI data array (128 bytes for HT20)
        
        Returns:
            int: Current buffer size (progress indicator)
        """
        if self._packet_count >= self.buffer_size:
            return self.buffer_size
        
        # Filter by expected CSI length (HT20: 128 bytes)
        if len(csi_data) != EXPECTED_CSI_LEN:
            self._filtered_count += 1
            if self._filtered_count % 50 == 1:
                print(f'[WARN] Filtered {self._filtered_count} packets with wrong SC count (got {len(csi_data)} bytes)')
            return self._packet_count
        
        # Initialize on first packet
        if not self._initialized:
            self._initialized = True
            print(f'P95: HT20 mode, {NUM_SUBCARRIERS} SC, guard [{GUARD_BAND_LOW}-{GUARD_BAND_HIGH}], DC={DC_SUBCARRIER}')
        
        # Store raw I/Q data (2 bytes per subcarrier)
        iq_data = bytearray(EXPECTED_CSI_LEN)
        for sc in range(NUM_SUBCARRIERS):
            i_idx = sc * 2
            q_idx = sc * 2 + 1
            iq_data[i_idx] = int(csi_data[i_idx]) & 0xFF
            iq_data[q_idx] = int(csi_data[q_idx]) & 0xFF
        
        self._file.write(iq_data)
        self._packet_count += 1
        
        if self._packet_count % 100 == 0:
            self._file.flush()
        
        return self._packet_count
    
    def _prepare_for_reading(self):
        """Close write mode and reopen for reading"""
        if self._file:
            self._file.flush()
            self._file.close()
            gc.collect()
        self._file = open(BUFFER_FILE, 'rb')
    
    def _read_all_packets(self):
        """Read all packets from file and pre-compute magnitudes"""
        self._file.seek(0)
        data = self._file.read()
        
        num_packets = len(data) // EXPECTED_CSI_LEN
        print(f"P95: File size {len(data)} bytes = {num_packets} packets")
        
        packets_mags = []
        for pkt_idx in range(num_packets):
            offset = pkt_idx * EXPECTED_CSI_LEN
            mags = []
            for sc in range(NUM_SUBCARRIERS):
                i_idx = offset + sc * 2
                q_idx = i_idx + 1
                # Espressif CSI format: [Imaginary, Real, ...] per subcarrier
                Q = data[i_idx]      # Imaginary first
                I = data[q_idx]      # Real second
                if I > 127:
                    I = I - 256
                if Q > 127:
                    Q = Q - 256
                mags.append(math.sqrt(I * I + Q * Q))
            packets_mags.append(mags)
            
            if (pkt_idx + 1) % 50 == 0:
                gc.collect()
        
        gc.collect()
        return packets_mags
    
    def _get_candidate_bands(self):
        """Generate valid candidate bands of 12 consecutive subcarriers."""
        candidates = []
        
        # Zone before DC (subcarriers 11-31)
        for start in range(GUARD_BAND_LOW, DC_SUBCARRIER - BAND_SIZE + 1):
            band = list(range(start, start + BAND_SIZE))
            if all(GUARD_BAND_LOW <= sc <= GUARD_BAND_HIGH and sc != DC_SUBCARRIER for sc in band):
                candidates.append(band)
        
        # Zone after DC (subcarriers 33-52)
        for start in range(DC_SUBCARRIER + 1, GUARD_BAND_HIGH - BAND_SIZE + 2):
            band = list(range(start, start + BAND_SIZE))
            if all(GUARD_BAND_LOW <= sc <= GUARD_BAND_HIGH and sc != DC_SUBCARRIER for sc in band):
                candidates.append(band)
        
        return candidates
    
    def _evaluate_band(self, packets_mags, band):
        """
        Evaluate a band by computing P95 and MV values.
        
        Note: Hampel filter is NOT applied during calibration. Outliers are useful
        information for band selection (P95 is already robust to outliers).
        Hampel is only applied during normal operation in the CSI processor.
        """
        # Calculate turbulence series
        turbulences = []
        for pkt_mags in packets_mags:
            turb = calculate_spatial_turbulence(pkt_mags, band)
            turbulences.append(turb)
        
        # Calculate moving variance series
        mv_series = calculate_moving_variance(turbulences, MVS_WINDOW_SIZE)
        del turbulences
        
        if not mv_series:
            return {'p95': float('inf'), 'mv_values': [], 'fp_estimate': 1.0}
        
        # P95 for band selection (fixed)
        p95 = calculate_percentile(mv_series, BAND_SELECTION_PERCENTILE)
        
        # Estimate FP rate
        fp_count = sum(1 for v in mv_series if v > MVS_THRESHOLD)
        fp_estimate = fp_count / len(mv_series)
        
        return {
            'p95': p95,
            'mv_values': mv_series,
            'fp_estimate': fp_estimate
        }
    
    def calibrate(self):
        """
        Calibrate by selecting the optimal band.
        
        Returns:
            tuple: (selected_band, mv_values) or (None, []) if failed
        """
        if self._packet_count < MVS_WINDOW_SIZE + 10:
            print("P95: Not enough packets for calibration")
            return None, []
        
        self._prepare_for_reading()
        
        packets = self._read_all_packets()
        if len(packets) < MVS_WINDOW_SIZE + 10:
            print("P95: Failed to read packets")
            return None, []
        
        print(f"P95: Analyzing {len(packets)} packets...")
        
        candidates = self._get_candidate_bands()
        
        if not candidates:
            print("P95: No valid candidate bands found")
            return None, []
        
        print(f"P95: Evaluating {len(candidates)} candidate bands...")
        
        # Evaluate each candidate
        band_results = []
        for i, band in enumerate(candidates):
            result = self._evaluate_band(packets, band)
            result['band'] = band
            band_results.append(result)
            
            if (i + 1) % 10 == 0:
                gc.collect()
                print(f"  Evaluating... {i+1}/{len(candidates)}")
        
        # Select band with best P95
        safe_margin = 0.15
        p95_limit = MVS_THRESHOLD - safe_margin
        
        valid_bands = [r for r in band_results if r['p95'] < p95_limit]
        
        if valid_bands:
            best_result = max(valid_bands, key=lambda r: r['p95'])
            print(f"P95: Found {len(valid_bands)} safe bands (P95 < {p95_limit:.2f})")
        else:
            print(f"P95: No bands with P95 < {p95_limit:.2f}, using lowest")
            best_result = min(band_results, key=lambda r: r['p95'])
        
        best_band = best_result['band']
        if best_band is None:
            print("P95: All candidates failed evaluation")
            return None, []
        
        mv_values = best_result['mv_values']
        
        # Report results
        print(f"P95: Band selection successful")
        print(f"  Selected: [{best_band[0]}-{best_band[-1]}]")
        print(f"  P95 MV: {best_result['p95']:.4f}")
        print(f"  Est. FP rate: {best_result['fp_estimate']*100:.1f}%")
        
        if self._filtered_count > 0:
            print(f"  Filtered: {self._filtered_count} packets (wrong SC count)")
        
        return best_band, mv_values
    
    def free_buffer(self):
        """Free resources after calibration"""
        if self._file:
            self._file.close()
            self._file = None
        
        try:
            os.remove(BUFFER_FILE)
        except OSError:
            pass
    
    def get_packet_count(self):
        """Get the number of packets currently in the buffer."""
        return self._packet_count
    
    def is_buffer_full(self):
        """Check if the buffer has collected enough packets."""
        return self._packet_count >= self.buffer_size
        
        gc.collect()


# Backward compatibility alias
BandCalibrator = P95Calibrator
